src/proposed/0140-budget-limit-threshold.md

```diff
@@ -0,0 +1,469 @@
+<!--
+  SPDX-License-Identifier: CC-BY-SA-4.0
+  Copyright 2023 UNSW
+-->
+
+# MCS: Budget limit thresholds on endpoints for SC Donation
+
+- Author: Mitchell Johnston
+- Proposed: 2023-08-15
+
+## Summary
+
+This is a two stage RFC that proposes to add a new `budget threshold' to
+endpoints, which would only allow scheduling context donation to occur if the
+SC's available budget exceeds the threshold. This would provide a mechanism to
+prevent budget-expiry from occurring in passive servers.
+
+As a second optional stage, this threshold value could also restrict the budget
+that a passive server is permitted to consume on a donated SC. We refer to this
+as a `budget limit'. This would help bound priority inversion in a system as
+well as reducing the trust that clients must place in passive servers they call.
+
+## Motivation
+
+### Thresholds
+
+This change regarding thresholds is motivated by a desire to prevent budget
+expiry from occurring inside passive servers, improving worst-case performance
+and response time analysis. Additionally, it would shift timeout exceptions into
+being a true error case, rather than an unavoidable, but routine occurrence on
+many systems. This is likely to simplify system design.
+
+Currently, there are almost no restrictions on the amount of budget required to
+enter a passive server. This almost inevitably leads to budget expiry, resulting
```

Comment:
nit: this almost inevitably leads to budget expiry
suggestion: this can lead to undesirably budget expiry in certain system designs

Not really invevitable, that's up to the system. If you have a system where overruns are tolerated, or where you think you have the WCET mostly sorted but not completely, you can have occasional budget expiry and handle it. Whether it's desirable or not is up to the system design and is a matter of policy. That said, *if* you wanted a system where you really did not want timeout exceptions,